File: examples/inProgress/CIL/programs/paper/dekker.c
/*
 * Dekker's algorithm
 */
// based on code found at http://jakob.engbloms.se/archives/65
// original file written by Jakob Engblom

#include <stdio.h>
#include <stdlib.h>

static volatile int flag1 = 0;
static volatile int flag2 = 0;
static volatile int turn  = 1;
//static volatile int gSharedCounter = 0;
//int gLoopCount = 5;

void critical1() {
	//gSharedCounter++;
}
void critical2() {
	//gSharedCounter++;
}

void dekker1(void) {
	flag1 = 1;
	turn = 2;
	while((flag2 == 1) && (turn == 2)) ;
	// Critical section
	critical1();
	// Let the other task run
	flag1 = 0;
}

void dekker2(void) {
	flag2 = 1;
	turn = 1;
	while((flag1 == 1) && (turn == 1)) ;
	// critical section
	critical2();
	// leave critical section
	flag2 = 0;
}

//
// Tasks, as a level of indirection
//
void task1(void) {
	// printf("Starting task1\n");
	// for (int i = 0; i < gLoopCount; i++) {
	while(1) {
		dekker1();
	}
}

void task2(void) {
	// printf("Starting task2\n");
	// for (int i = 0; i < gLoopCount; i++) {
	while(1) {
		dekker2();
	}
}

int main(void) {
	//int expected_sum = gLoopCount * 2;

	/* Start the threads */
	spawn(task1);
	spawn(task2);

	/* Wait for the threads to end */
	sync();
	
	//printf("Computed %d\n", gSharedCounter);
	//return gSharedCounter;
	return 0;
}

/*
--- START MODEL-CHECKING
--- mod MODEL-CHECKER
--- including C-program-dekker .
	including PL-MODEL-CHECKER .
	op state : Bag -> Model-Checker-State .
	op Start : -> Model-Checker-State .
	---eq Start = state(eval('program-dekker)) .
	op enabled : Id -> Prop .
	eq
		state(? < T >... < k > 'Apply('Closure(X:Id,, ?,, ?),, ?) ...</ k > ...</ T >) |= enabled(X:Id) = true .
--- endm
--- END MODEL-CHECKING

erew modelCheck(Start, <> enabled) .

*/

File: examples/languages/educational/simple/untyped/dekker.m
load simple-untyped-compiled
load ../../../k-model-checker
select SIMPLE-UNTYPED .


mod MODEL-CHECK-TEST is
	including SIMPLE-UNTYPED .
	including PL-MODEL-CHECKER .
	op modelBag : -> Bag .
	eq modelBag = run('pDekker) .
	op Start : -> Model-Checker-State .
	op state : Bag -> Model-Checker-State .
	eq Start = state(modelBag) .
	
	op set : Id -> Prop .
	--- eq state(
		--- B0:Bag
		--- < T > 
			--- B1:Bag
			--- < genv > 
				--- M0:Map
				--- Id X:Id(.List{K}) |-> Loc:K
			--- </ genv >
			--- < mem > 
				--- M1:Map  
				--- Loc:K |-> Rat piece(1, 8)(.List{K})
			--- </ mem >
		--- </ T >	
	--- ) |= set(X:Id) = true .
	
	op enabled : Id -> Prop .
	eq state(
		B0:Bag
		< T > 
			B1:Bag
			< threads >
				B2:Bag
				< thread >
					B3:Bag
					< k > 'Apply('Closure(Id X:Id(.List{K}),, K0:K,, K1:K),, K3:K) ~> K:K </ k >
				</ thread >
			</ threads >
		</ T >
	) |= enabled(X:Id) = true .
	--- 'Apply('Closure(Id critical2(.List{K}),, K0:K,, K1:K),, K3:K) ~> K:K
	
	--- op finishWith : Nat -> Prop .
	--- eq state(
		--- B0:Bag
		--- < resultValue > 'tv(Rat N:Nat(.List{K}),, Base-Type int(.List{K})) </ resultValue >	
	--- ) |= finishWith(N:Nat) = true .
endm

--- rew run('pDekker) .



red modelCheck(Start, [] ~ (enabled(critical1) /\ enabled(critical2))) .

---search modelBag =>! B:Bag . --- mutual exclusion and no deadlock
--- red modelCheck(Start, <> (set(done1) /\ set(done2))) .
--- red modelCheck(Start, <> (set(done1) )) .
--- red modelCheck(Start, <> enabled(critical)) .
 ---red modelCheck(Start, <> enabled(dekker1)) .
--- red modelCheck(Start, [] ~ (enabled(dekker1) /\ enabled(dekker2))) .
---red modelCheck(Start, (<> enabled(task1) /\ <> enabled(task2))) .
--- red modelCheck(Start, <> finishWith(10)) .
---red modelCheck(Start, (<> enabled(dekker1) /\ <> enabled(dekker2)) -> (<> enabled(critical1) /\ <> enabled(critical2))) .
